tests/apps/tracingapp/app.go

```diff
@@ -0,0 +1,275 @@
+)
+
+var (
+	appPort      = 3000
+	daprHTTPPort = 3500
+
+	httpClient = utils.NewHTTPClient()
+
+	httpMethods []string
+
+	zipkinEndpoint = "http://dapr-zipkin:9411"
+	serviceName    = "tracingapp"
+
+	tracer      trace.Tracer
+	propagators propagation.TextMapPropagator
+)
+
+const (
+	jsonContentType = "application/json"
+	zipkinSpans     = "/api/v2/spans"
+	zipkinTraces    = "/api/v2/traces"
+
+	expectedDaprChildSpanNameTemplate = "calllocal/%s/invoke/something"
+)
+
+func init() {
+	p := os.Getenv("DAPR_HTTP_PORT")
+	if p != "" && p != "0" {
+		daprHTTPPort, _ = strconv.Atoi(p)
+	}
+	p = os.Getenv("PORT")
+	if p != "" && p != "0" {
+		appPort, _ = strconv.Atoi(p)
+	}
+	p = os.Getenv("ZIPKIN_ENDPOINT")
+	if p != "" {
+		zipkinEndpoint = p
+	}
+	p = os.Getenv("SERVICE_NAME")
+	if p != "" {
+		serviceName = p
+	}
+}
+
+type appResponse struct {
+	SpanName *string `json:"spanName,omitempty"`
+	Message  string  `json:"message,omitempty"`
+}
+
+// indexHandler is the handler for root path
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	log.Println("indexHandler is called")
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(appResponse{Message: "OK"})
+}
+
+// triggerInvoke is the handler for end-to-end to start the invoke stack
+func triggerInvoke(w http.ResponseWriter, r *http.Request) {
+	log.Printf("Processing %s %s", r.Method, r.URL.RequestURI())
+	uuidObj, _ := uuid.NewRandom()
+	uuid := uuidObj.String()
+	ctx := propagators.Extract(r.Context(), propagation.HeaderCarrier(r.Header))
+	newCtx, span := tracer.Start(ctx, uuid)
+	defer span.End()
+
+	query := r.URL.Query()
+	appId := query.Get("appId")
+
+	url := fmt.Sprintf("http://127.0.0.1:%d/v1.0/invoke/%s/method/invoke/something", daprHTTPPort, appId)
+	/* #nosec */
+	req, _ := http.NewRequest(http.MethodPost, url, nil)
+	req = req.WithContext(newCtx)
+	hc := propagation.HeaderCarrier(req.Header)
+	propagators.Inject(newCtx, hc)
+	req.Header.Add("Content-Type", jsonContentType)
+
+	fmt.Printf("span's name is %s and invoke url is %s\n", uuid, url)
+
+	res, err := httpClient.Do(req)
+	if err != nil {
+		log.Println(err)
+
+		w.WriteHeader(http.StatusInternalServerError)
+		json.NewEncoder(w).Encode(appResponse{
+			SpanName: &uuid,
+			Message:  err.Error(),
+		})
+	}
+
+	if res.StatusCode != http.StatusOK {
+		w.WriteHeader(http.StatusExpectationFailed)
+		json.NewEncoder(w).Encode(appResponse{
+			SpanName: &uuid,
+			Message:  fmt.Sprintf("expected status code %d, got %d", http.StatusOK, res.StatusCode),
+		})
+	}
+
+	json.NewEncoder(w).Encode(appResponse{
+		SpanName: &uuid,
+		Message:  "OK",
+	})
+}
+
+// invoke is the handler for end-to-end to invoke
+func invoke(w http.ResponseWriter, r *http.Request) {
+	log.Printf("Processing %s %s", r.Method, r.URL.RequestURI())
+	ctx := propagators.Extract(r.Context(), propagation.HeaderCarrier(r.Header))
+	_, span := tracer.Start(ctx, "invokedMethod")
+	defer span.End()
+
+	// Pretend to do work
+	time.Sleep(time.Second)
```

Comment:
Missed a chance to actually mine some bitcoins here 🙃

Reply:
It would be interesting to see the cost-benefit analysis of doing so in the cloud when you pay for reserved capacity and not for utilized capacity. If you have reserved and unused capacity going to waste, then mining might make sense. Good idea for our LH tests.